src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,382 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import number_theory.legendre_symbol.quadratic_reciprocity
+import data.zmod.coprime
+
+/-!
+# The Jacobi Symbol
+
+We define the Jacobi symbol and prove its main properties.
+
+## Main definitions
+
+We define the Jacobi symbol, `jacobi_sym a b`, for integers `a` and natural numbers `b`
+as the product over the prime factors `p` of `b` of the Legendre symbols `zmod.legendre_sym p a`.
+This agrees with the mathematical definition when `b` is odd.
+
+The prime factors are obtained via `nat.factors`. Since `nat.factors 0 = []`,
+this implies in particular that `jacobi_sym a 0 = 1` for all `a`.
+
+## Main statements
+
+We prove the main properties of the Jacobi symbol, including the following.
+
+* Multiplicativity in both arguments (`jacobi_sym_mul_left`, `jacobi_sym_mul_right`)
+
+* The value of the symbol is `1` or `-1` when the arguments are coprime
+  (`jacobi_sym_eq_one_or_neg_one`)
+
+* The symbol vanishes if and only if `b ≠ 0` and the arguments are not coprime
+  (`jacobi_sym_eq_zero_iff`)
+
+* If the symbol has the value `-1`, then `a : zmod b` is not a square
+  (`nonsquare_of_jacobi_sym_eq_neg_one`); the converse holds when `b = p` is a prime
+  (`nonsquare_iff_jacobi_sym_eq_neg_one`); in particular, in this case `a` is a
+  square mod `p` when the symbol has the value `1` (`is_square_of_jacobi_sym_eq_one`).
+
+* Quadratic reciprocity (`jacobi_sym_quadratic_reciprocity`,
+  `jacobi_sym_quadratic_reciprocity_one_mod_four`,
+  `jacobi_sym_quadratic_reciprocity_three_mod_four`)
+
+* The supplementary laws for `a = -1`, `a = 2`, `a = -2` (`jacobi_sym_neg_one`, `jacobi_sym_two`,
+  `jacobi_sym_neg_two`)
+
+* The symbol depends on `a` only via its residue class mod `b` (`jacobi_sym_mod_left`)
+  and on `b` only via its residue class mod `4*a` (`jacobi_sym_mod_right`)
+
+## Notations
+
+We define the notation `J(a | b)` for `jacobi_sym a b`, localized to `number_theory_symbols`.
+
+## Tags
+Jacobi symbol, quadratic reciprocity
+-/
+
+section jacobi
+
+/-!
+### Definition of the Jacobi symbol
+
+We define the Jacobi symbol $\Bigl\frac{a}{b}\Bigr$ for integers `a` and natural numbers `b` as the
+product of the Legendre symbols $\Bigl\frac{a}{p}\Bigr$, where `p` runs through the prime divisors
+(with multiplicity) of `b`, as provided by `b.factors`. This agrees with the Jacobi symbol
+when `b` is odd and gives less meaningful values when it is not (e.g., the symbol is `1`
+when `b = 0`). This is called `jacobi_sym a b`.
+
+We define localized notation (locale `number_theory_symbols`) `J(a | b)` for the Jacobi
+symbol `jacobi_sym a b`. (Unfortunately, there is no subscript "J" in unicode.)
+-/
+
+open zmod nat
```

Comment:
Legendre symbol is called [zmod.legendre_sym](https://leanprover-community.github.io/mathlib_docs/number_theory/legendre_symbol/quadratic_reciprocity.html#zmod.legendre_sym), we should have `zmod` as a namespace here, for coherence.

If you do this, don't forget to update the names in the docstring above, so the hyperlink will work.

Reply:
Actually, I was thinking of putting `legendre_sym` in the root namespace instead (and use `legendre_sym.xxx` in place of `legendre_sym_xxx`). Do you think this would be reasonable? (The `zmod` namespace was what I found when I started working on the Legendre symbol; it goes back to the original formalization by Chris Hughes I assume.)

But this should be done in a separate PR.

Reply:
Can you add the namespace to this one? It should be just a one line change, and then remove both of them in another PR. Thanks!

Reply:
Ignore the doc. The rest should be automatic, if you're in the namespace `zmod` then it is opened. It's enough to add `namespace zmod`, for example at line 75, and close it at the end. If I am missing something and it causes much trouble don't worry, we will fix it later.

Reply:
OK, done.